presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
     {
         checkInputs(fragment.getRemoteSourceNodes(), rddInputs, broadcastInputs);
 
-        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
-        verify(tableScans.isEmpty(), "no table scans is expected");
-
-        PrestoSparkTaskDescriptor taskDescriptor = createIntermediateTaskDescriptor(session, tableWriteInfo, fragment);
-        SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(taskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
-
-        if (rddInputs.size() == 0) {
-            checkArgument(fragment.getPartitioning().equals(SINGLE_DISTRIBUTION), "SINGLE_DISTRIBUTION partitioning is expected: %s", fragment.getPartitioning());
-            return sparkContext.parallelize(ImmutableList.of(serializedTaskDescriptor), 1)
-                    .mapPartitionsToPair(createTaskProcessor(
-                            executorFactoryProvider,
-                            taskStatsCollector,
-                            toTaskProcessorBroadcastInputs(broadcastInputs)));
-        }
+        PrestoSparkTaskDescriptor taskDescriptor = new PrestoSparkTaskDescriptor(
+                session.toSessionRepresentation(),
+                session.getIdentity().getExtraCredentials(),
+                fragment,
+                tableWriteInfo);
+        SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(
+                taskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
 
+        Optional<Integer> numberOfInputPartitions = Optional.empty();
```

Comment:
Curious what does this `numberOfInputPartitions` mean? -- Looks like it's decided by the partition number of other RDDs?

Reply:
It represents number of partitions that are expected by the rdd based on it's inputs. The number of partitions should be equal for all input rdds. If there's no shuffle inputs we don't have to create empty partitions if the splits are missing for some buckets. Think of it as of an optimization for a query like

```
select col, count(*)
from table_bucketed_by_col
group by col
```

If some buckets are missing we are free to not to create tasks for buckets that are missing.

But for a query like 

```
select col, count(*)
from table_bucketed_by_col t1
RIGHT OUTER JOIN non_bucketed_table t2
WHERE t1.col = t2.col
```

even if some buckets are missing for t1 we still need to create a task for correctness

Reply:
@arhimondr : I see. So essentially `numberOfInputPartitions` comes from the left-most join table right ?

i.e. considering 
```
            JOIN
          /   |   \
        A    B    C
```

Once `rddA` is created, the number of partitions in `rddA` will be populated to `B` and `C` for the number of partitions?

Reply:
> I see. So essentially numberOfInputPartitions comes from the left-most join table right ?

No. It should be determined by the partitions produced by shuffle.

```
           JOIN
          /   |   \
        A    B    C
```

If table A and C are bucketed, but table B is non bucketed - we will have to shuffle table B. Thus the number of partitions will be present, and will be set to the number of partitions produced by the shuffle over B.

If all tables A, B and C are bucketed, and if bucket 0 is missing for all of them - there's no need to schedule a task for that bucket. This optimization cannot be done if at least one table is non bucketed, as there's no way to know in advance if there will be any rows for that bucket produced by the shuffle.

Reply:
@arhimondr : Maybe add some comments about the intention of`expectedNumberOfPartitions` (used to eliminate tasks when there are missing buckets), and its behavior when there are both bucketed table input and shuffle input, since it's not intuitive at first glance :)